anndata/_io/h5ad.py

```diff
@@ -87,8 +87,14 @@ def write_h5ad(
         adata.strings_to_categoricals(adata.raw.var)
     dataset_kwargs = {**dataset_kwargs, **kwargs}
     filepath = Path(filepath)
-    mode = "a" if adata.isbacked else "w"
-    if adata.isbacked:  # close so that we can reopen below
+    copy = (
+        not adata.isbacked
+        or not filepath.exists()
+        or not filepath.samefile(adata.filename)
+    )
+
+    mode = "a" if not copy else "w"
+    if not copy:  # close so that we can reopen below
```

Comment:
I'm not sure this is necessary. What if we just didn't call `adata.file.open(filepath, "r+")` at the bottom?